tokio/src/sync/mpsc/bounded.rs

```diff
@@ -178,8 +184,18 @@ impl<T> Receiver<T> {
     /// }
     /// ```
     pub async fn recv(&mut self) -> Option<T> {
-        use crate::future::poll_fn;
-        poll_fn(|cx| self.chan.recv(cx)).await
+        use crate::{future::poll_fn, sync::mpsc::chan::Semaphore};
+        let res = poll_fn(|cx| self.chan.recv(cx)).await;
+
+        // If there is already an underflow, the permit released by recv is
+        // cancelled and the underflow is reduced accordingly.
+        if self.underflow > 0 {
+            self.chan.semaphore().reduce_permits(1);
+
+            self.underflow -= 1;
+        }
+
+        res
```

Comment:
This wont work. A sender may have consumed the permit before you get to the `reduce_permits` call, in which case the underflow is lost.

Reply:
This did indeed open the door to race conditions, thank you. I added a parameter that allows the `recv` method to conditionally release a permit. This way, no sender can acquire a permit while the channel is overflowing. What do you think about this approach?